controllers/storagecluster/resourceprofile.go

```diff
@@ -0,0 +1,199 @@
+package storagecluster
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+	"time"
+
+	ocsv1 "github.com/red-hat-storage/ocs-operator/v4/api/v1"
+	"github.com/red-hat-storage/ocs-operator/v4/controllers/defaults"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var (
+	lastResourceProfile              string
+	resourceProfileChangeisHappening bool
+	resourceProfileChangeStartTime   time.Time
+	resourcePrfoileChangeTimeout     = 30 * time.Minute
+)
+
+const (
+	resourceProfileChangeApplying = "New resource profile is being applied"
+	resourceProfileChangeFailed   = "New resource profile failed to apply, please revert to the last working resource profile"
+)
+
+func (r *StorageClusterReconciler) ensureResourceProfileChangeApplied(sc *ocsv1.StorageCluster) error {
```

Comment:
Why do we need to verify every detail of the resources applied? Can't we just confirm that the reconcile is completed (the CephCluster CR completion status), and perhaps as one extra step confirm that no ceph pods are pending? We should trust that the rook operator will do its job. A simple pod query with the labelSelector `rook_cluster=<namespace>` may be sufficient, then check no pods are in pending status.

Reply:
Tried this, Here are the issues

1. When the cephcluster spec is updated it takes a few seconds for rook to mark it as progressing, So in this brief moment both our checks pass i. Cephcluster being Ready, ii, None of the pods with the label are pending. Hence ocs operator thinks the resource profile has been applied successfully. While the truth is nothing has started out yet.

2. Even if we bypass the above problem somehow when rook starts processing the changes, The pod restarts are done one by one and in between two restarts there are brief moments when the cephcluster reports as ready & non of the pods with the label are pending. So OCS can again take it as a false positive & report the new profile has been applied already when it's actually in the halfway stage.

3. When we are trying to do a get on the pods with the label rook_cluster=<namespace> we are also getting crashcollectors, exporters, osd-prepare pods which are not actually being affected by this profile mechanism & if they are pending due to some other reason ocs will still report that it failed to apply the resource profile while it's not the case.

Reply:
Here is another approach... The CephCluster CR allows setting [labels](https://github.com/rook/rook/blob/master/deploy/examples/cluster.yaml#L202-L208) for different types of daemons. What if you set the name of the profile as a label for `all` components? For example, a label for `odf-profile=lean|balanced|performance`. Then all you need is a single query to see if daemons have the expected profile.

Reply:
Tried it, Using the labels looks like a better idea than comparing the ResourceRequirements. Have updated the code. But, instead of all have given labels to specifically mon, mgr, osd on the cephcluster CR & mds on the filesystem CR, rgw on the objectstore CR.